```
error[E0050]: method `register_webxr` has 4 parameters but the declaration in trait `compositing::windowing::EmbedderMethods::register_webxr` has 5
  --> ports/glutin/embedder.rs:94:9
   |
94 | /         &mut self,
95 | |         xr: &mut webxr::MainThreadRegistry,
96 | |         _executor: WebGlExecutor,
97 | |         _surface_provider_registration: SurfaceProviders
   | |________________________________________________________^ expected 5 parameters, found 4
   |
   = note: `register_webxr` from trait: `fn(&mut Self, &mut webxr_api::registry::MainThreadRegistry<surfman_chains::SwapChains<webxr_api::SwapChainId>>, crossbeam_channel::channel::Sender<std::boxed::Box<(dyn std::ops::FnOnce() + std::marker::Send + 'static)>>, std::sync::Arc<std::sync::Mutex<std::collections::HashMap<webxr_api::session::SessionId, std::boxed::Box<(dyn surfman_chains::SurfaceProvider + std::marker::Send + 'static)>>>>, embedder_traits::EmbedderProxy)`
error: aborting due to previous error
```

--- a/components/compositing/windowing.rs
+++ b/components/compositing/windowing.rs
@@ -5,7 +5,7 @@
 //! Abstract windowing methods. The concrete implementations of these can be found in `platform/`.
 
 use canvas::{SurfaceProviders, WebGlExecutor};
-use embedder_traits::EventLoopWaker;
+use embedder_traits::{EmbedderProxy, EventLoopWaker};
 use euclid::Scale;
 #[cfg(feature = "gl")]
 use gleam::gl;
@@ -190,6 +190,7 @@ pub trait EmbedderMethods {
         _: &mut webxr::MainThreadRegistry,
         _: WebGlExecutor,
         _: SurfaceProviders,
+        _: EmbedderProxy,
     ) {
     }
 }

--- a/components/servo/lib.rs
+++ b/components/servo/lib.rs
@@ -481,6 +481,7 @@ where
                     &mut webxr_main_thread,
                     webgl_executor,
                     webxr_surface_providers,
+                    embedder_proxy.clone(),
                 );
             }
         }
@@ -521,7 +522,7 @@ where
         let (constellation_chan, sw_senders) = create_constellation(
             opts.user_agent.clone(),
             opts.config_dir.clone(),
-            embedder_proxy.clone(),
+            embedder_proxy,
             compositor_proxy.clone(),
             time_profiler_chan.clone(),
             mem_profiler_chan.clone(),

--- a/ports/glutin/embedder.rs
+++ b/ports/glutin/embedder.rs
@@ -14,7 +14,7 @@ use glutin::EventsLoopClosed;
 use rust_webvr::GlWindowVRService;
 use servo::canvas::{SurfaceProviders, WebGlExecutor};
 use servo::compositing::windowing::EmbedderMethods;
-use servo::embedder_traits::EventLoopWaker;
+use servo::embedder_traits::{EmbedderProxy, EventLoopWaker};
 use servo::servo_config::{opts, pref};
 use servo::webvr::VRServiceManager;
 use servo::webvr_traits::WebVRMainThreadHeartbeat;
@@ -94,7 +94,8 @@ impl EmbedderMethods for EmbedderCallbacks {
         &mut self,
         xr: &mut webxr::MainThreadRegistry,
         _executor: WebGlExecutor,
-        _surface_provider_registration: SurfaceProviders
+        _surface_provider_registration: SurfaceProviders,
+        _embedder_proxy: EmbedderProxy,
     ) {
         if pref!(dom.webxr.test) {
             xr.register_mock(webxr::headless::HeadlessMockDiscovery::new());

--- a/ports/libsimpleservo/api/src/lib.rs
+++ b/ports/libsimpleservo/api/src/lib.rs
@@ -20,7 +20,9 @@ use servo::compositing::windowing::{
     WindowMethods,
 };
 use servo::embedder_traits::resources::{self, Resource, ResourceReaderMethods};
-use servo::embedder_traits::{EmbedderMsg, MediaSessionEvent, PromptDefinition, PromptOrigin};
+use servo::embedder_traits::{
+    EmbedderMsg, EmbedderProxy, MediaSessionEvent, PromptDefinition, PromptOrigin,
+};
 use servo::euclid::{Point2D, Rect, Scale, Size2D, Vector2D};
 use servo::keyboard_types::{Key, KeyState, KeyboardEvent};
 use servo::msg::constellation_msg::TraversalDirection;
@@ -764,7 +766,10 @@ impl EmbedderMethods for ServoEmbedderCallbacks {
         registry: &mut webxr::MainThreadRegistry,
         executor: WebGlExecutor,
         surface_providers: SurfaceProviders,
+        embedder_proxy: EmbedderProxy,
     ) {
+        use ipc_channel::ipc::{self, IpcReceiver};
+        use webxr::openxr;
         debug!("EmbedderMethods::register_xr");
         assert!(
             self.xr_discovery.is_none(),
@@ -772,12 +777,50 @@ impl EmbedderMethods for ServoEmbedderCallbacks {
         );
 
         struct ProviderRegistration(SurfaceProviders);
-        impl webxr::openxr::SurfaceProviderRegistration for ProviderRegistration {
+        impl openxr::SurfaceProviderRegistration for ProviderRegistration {
             fn register(&self, id: webxr_api::SessionId, provider: servo::canvas::SurfaceProvider) {
                 self.0.lock().unwrap().insert(id, provider);
             }
-            fn clone(&self) -> Box<dyn webxr::openxr::SurfaceProviderRegistration> {
+            fn clone(&self) -> Box<dyn openxr::SurfaceProviderRegistration> {
                 Box::new(ProviderRegistration(self.0.clone()))
+            }
+        }
+
+        #[derive(Clone)]
+        struct ContextMenuCallback(EmbedderProxy);
+
+        struct ContextMenuFuture(IpcReceiver<ContextMenuResult>);
+
+        impl openxr::ContextMenuProvider for ContextMenuCallback {
+            fn open_context_menu(&self) -> Box<dyn openxr::ContextMenuFuture> {
+                let (sender, receiver) = ipc::channel().unwrap();
+                self.0.send((
+                    None,
+                    EmbedderMsg::ShowContextMenu(
+                        sender,
+                        Some("Would you like to exit the XR session?".into()),
+                        vec!["Exit".into()],
+                    ),
+                ));
+
+                Box::new(ContextMenuFuture(receiver))
+            }
+            fn clone_object(&self) -> Box<dyn openxr::ContextMenuProvider> {
+                Box::new(self.clone())
+            }
+        }
+
+        impl openxr::ContextMenuFuture for ContextMenuFuture {
+            fn poll(&self) -> openxr::ContextMenuResult {
+                if let Ok(result) = self.0.try_recv() {
+                    if let ContextMenuResult::Selected(0) = result {
+                        openxr::ContextMenuResult::ExitSession
+                    } else {
+                        openxr::ContextMenuResult::Dismissed
+                    }
+                } else {
+                    openxr::ContextMenuResult::Pending
+                }
             }
         }
 
@@ -794,6 +837,7 @@ impl EmbedderMethods for ServoEmbedderCallbacks {
         let discovery = webxr::openxr::OpenXrDiscovery::new(
             Box::new(GlThread(executor)),
             Box::new(ProviderRegistration(surface_providers)),
+            Box::new(ContextMenuCallback(embedder_proxy)),
         );
         registry.register(discovery);
     }
@@ -804,6 +848,7 @@ impl EmbedderMethods for ServoEmbedderCallbacks {
         registry: &mut webxr::MainThreadRegistry,
         _executor: WebGlExecutor,
         _surface_provider_registration: SurfaceProviders,
+        _embedder_proxy: EmbedderProxy,
     ) {
         debug!("EmbedderMethods::register_xr");
         if let Some(discovery) = self.xr_discovery.take() {

